geth/jail/jail_cell_test.go

```diff
@@ -109,3 +107,190 @@ func (s *JailTestSuite) TestJailLoopInCall() {
+
+	err = cell.Set("__captureResponse", func(val string) otto.Value {
+		go func() { items <- val }()
+		return otto.UndefinedValue()
+	})
+	require.NoError(err)
+
+	_, err = cell.Run(`
+		function callRunner(namespace){
+			return setTimeout(function(){
+				__captureResponse(namespace);
+			}, 1000);
+		}
+	`)
+	require.NoError(err)
+
+	for i := 0; i < 100; i++ {
+		_, err = cell.Call("callRunner", nil, "softball")
+		require.NoError(err)
+	}
+
+	for i := 0; i < 100; i++ {
+		select {
+		case received := <-items:
+			require.Equal(received, "softball")
+			break
+		case <-time.After(5 * time.Second):
+			require.Fail("test timed out")
+		}
+	}
+}
+
+func (s *JailTestSuite) TestJailFetchPromise() {
+	body := `{"key": "value"}`
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Add("Content-Type", "application/json")
+		w.Write([]byte(body))
+	}))
+	defer server.Close()
+
+	require := s.Require()
+	require.NotNil(s.jail)
+
+	cell, err := s.jail.NewCell(testChatID)
+	require.NoError(err)
+	require.NotNil(cell)
+
+	dataCh := make(chan otto.Value, 1)
+	errCh := make(chan otto.Value, 1)
+
+	err = cell.Set("__captureSuccess", func(res otto.Value) { dataCh <- res })
+	require.NoError(err)
+	err = cell.Set("__captureError", func(res otto.Value) { errCh <- res })
+	require.NoError(err)
+
+	// run JS code for fetching valid URL
+	_, err = cell.Run(`fetch('` + server.URL + `').then(function(r) {
+		return r.text()
+	}).then(function(data) {
+		__captureSuccess(data)
+	}).catch(function (e) {
+		__captureError(e)
+	})`)
+	require.NoError(err)
+
+	select {
+	case data := <-dataCh:
+		require.True(data.IsString())
+		require.Equal(body, data.String())
+	case err := <-errCh:
+		require.Fail("request failed", err)
+	case <-time.After(1 * time.Second):
+		require.Fail("test timed out")
+	}
+}
+
+func (s *JailTestSuite) TestJailFetchCatch() {
+	require := s.Require()
+	require.NotNil(s.jail)
+
+	cell, err := s.jail.NewCell(testChatID)
+	require.NoError(err)
+	require.NotNil(cell)
+
+	dataCh := make(chan otto.Value, 1)
+	errCh := make(chan otto.Value, 1)
+
+	err = cell.Set("__captureSuccess", func(res otto.Value) { dataCh <- res })
+	require.NoError(err)
+	err = cell.Set("__captureError", func(res otto.Value) { errCh <- res })
+	require.NoError(err)
+
+	// run JS code for fetching invalid URL
+	_, err = cell.Run(`fetch('http://👽/nonexistent').then(function(r) {
+		return r.text()
+	}).then(function(data) {
+		__captureSuccess(data)
+	}).catch(function (e) {
+		__captureError(e)
+	})`)
+	require.NoError(err)
+
+	select {
+	case data := <-dataCh:
+		require.Fail("request should have failed, but returned", data)
+	case e := <-errCh:
+		require.True(e.IsObject())
+		name, err := e.Object().Get("name")
+		require.NoError(err)
+		require.Equal("Error", name.String())
+		_, err = e.Object().Get("message")
+		require.NoError(err)
+	case <-time.After(1 * time.Second):
+		require.Fail("test timed out")
+	}
+}
```

Comment:
The same comments about this and the next test.

Reply:
Sorry, didn't get it. What's the issue here?

Reply:
Ok, got it. I thought this comment is related to timer.